libportal/portal.c

```diff
@@ -259,7 +259,27 @@ xdp_portal_init (XdpPortal *portal)
 {
   int i;
 
-  portal->bus = g_bus_get_sync (G_BUS_TYPE_SESSION, NULL, NULL);
+  /* g_bus_get_sync() returns a singleton. In the test suite we may restart
+   * the session bus, so we have to manually connect to the new bus */
+  if (getenv("LIBPORTAL_TEST_SUITE"))
+    {
+      const char *addr = getenv ("DBUS_SESSION_BUS_ADDRESS");
+      g_autoptr(GError) error = NULL;
+
+      if (addr == NULL)
+        abort();
+
+      portal->bus = g_dbus_connection_new_for_address_sync (addr,
+                                                            G_DBUS_CONNECTION_FLAGS_AUTHENTICATION_CLIENT|G_DBUS_CONNECTION_FLAGS_MESSAGE_BUS_CONNECTION,
+                                                            NULL, NULL, &error);
+      if (error)
+        {
+          g_error("Failed to connect to bus: %s\n", error->message);
+          abort();
+        }
+    }
+  else
+    portal->bus = g_bus_get_sync (G_BUS_TYPE_SESSION, NULL, NULL);
```

Comment:
I think this would look a bit less convoluted if you split the first part into a separate function, something like this (notice the subtle code style suggestions too!):

```c
static GDBusConnection *
create_bus_from_address (const char  *address,
                         GError     **error)
{
  g_autoptr(GDBusConnection) bus = NULL;

  if (!address)
    {
      // Please think of a better message :)
      g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED, "Invalid D-Bus address");
      return NULL;
    }

  bus = g_dbus_connection_new_for_address_sync (address,
                                                G_DBUS_CONNECTION_FLAGS_AUTHENTICATION_CLIENT |
                                                G_DBUS_CONNECTION_FLAGS_MESSAGE_BUS_CONNECTION,
                                                NULL, NULL,
                                                &error);

  return g_steal_pointer (&bus);
}

...

  if (getenv ("LIBPORTAL_TEST_SUITE"))
    portal->bus = create_bus_from_address (getenv ("DBUS_SESSION_BUS_ADDRESS", &error);
  else
    portal->bus = g_bus_get_sync (G_BUS_TYPE_SESSION, NULL, &error);

  if (error)
    {
      g_critical ("Could not create XdpPortal instance: %s", error->message);
      abort ();
    }

  g_assert (portal->bus != NULL);
```

Reply:
Done, thanks. Though I admit my error message is not exactly the pinnacle of creativity either...